keps/sig-network/2030-endpointslice-subsetting/README.md

```diff
@@ -0,0 +1,212 @@
+
+## Motivation
+
+There has been some discussion that the Alpha Service Topology API requires the
+user to do too much. Users generally want the same behavior for all services
+where topology is concerned: avoid cross-zone traffic when in-zone endpoints are
+available and have enough capacity. This problem is exacerbated by the
+introduction of Multi-Cluster services where cross-region services become a
+reality and some regions may be more desirable than others as failover
+locations.
+
+As clusters and services grow, we’re also seeing that the current proxy
+implementations are stretched to their limits as all endpoints for the entire
+cluster are tracked independently by each node.
+
+The best user experience seems like it would be to have the platform - with a
+first class controller and/or a provider specific implementation - intelligently
+prioritize endpoints based on topology, capacity, and any other useful cost
+metrics to aid in traffic shaping, or just to reduce the amount of global
+resource tracking required by every node. However, we are currently missing a
+way to allow endpoints to be targeted to specific subsets of nodes.
+
+### Goals
+
+- Provide the building blocks to allow EndpointSlices to target specific subsets
+  of nodes.
+- EndpointSlices subsetting will be fully backwards compatible for older
+  consumers of the EndpointSlice API.
+- Design is flexible enough for multiple implementations and experimentation.
+- Minimal duplication of data.
+- Room for future enhancements, for example weighted endpoints or Slices.
+
+### Non-Goals
+
+- Define how subsetting should be used.
+- Design the controller responsible for subsetting endpoints.
+- An API for telling controllers how a service should be subsetted.
+
+Many of these are being tackled by the follow up [KEP
+#2004](https://github.com/kubernetes/enhancements/issues/2004).
+
+## Proposal
+
+Two new topology based labels will be introduced for EndpointSlices to support
+subsetting:
+
+```
+endpointslice.kubernetes.io/for-zone
+endpointslice.kubernetes.io/for-region
+```
+
+In the future this pattern may be expanded to include other concepts or
+topologies. A simple pattern like this will allow EndpointSlices to be delivered
+to consumers in a specific zone or region.
+
+
+### Risks and Mitigations
+
+This approach does not allow a single EndpointSlice to target multiple zones or
+regions. Any approach that enabled that would be significantly more complicated.
+The initial proposal in [KEP
+#2004](https://github.com/kubernetes/enhancements/issues/2004) suggests that
+this won't be an issue. 
+
+## Design Details
+
+### API Consumption
+
+EndpointSlice API consumers can use the following selectors to consume
+EndpointSlices:
+
+#### Select EndpointSlices Without Labels for Backwards Compatibility
+```
+matchExpressions:
+  - {key: endpointslice.kubernetes.io/for-zone, operator: DoesNotExist}
+  - {key: endpointslice.kubernetes.io/for-region, operator: DoesNotExist}
+```
```

Comment:
Does it mean the EndpointSliceController will be producing both the EndpointSlices without labels and with labels?

Why can't we assume that for backward-compatibility, we simply watch for *all* endpoint slices?

Reply:
Part of the appeal of this approach was that every EndpointSlice update would not have to be delivered to every node. With that said, it seems like the overhead of multiple watches might actually be worse than just client side filtering in kube-proxy. Do you have any data as far as what would be preferable here?

Reply:
I completely understand and support the motivation (we've been talking about that originally when designing EndpointSlices but there were different issues that weren't easy to solve and we didn't want to block the whole effort (I just shared you internally a braindump doc I wrote back then around this topic).

That said, I'm not yet convinced I understand how this can work (it doesn't mean the approach doesn't work, it means for now that I don't understand it) so trying to understand how you would envision the implementing, and whether instead of reducing amount of watches we won't end up with the need of writing N objects instead of 1.
So a sketch of the potential changes in the implementation would be super useful.

Reply:
This KEP is intentionally small since it's really just describing how kube-proxy/other consumers would consume EndpointSlices that had been subset. I think the more interesting KEP PR, complete with diagrams, is https://github.com/kubernetes/enhancements/pull/2005. 

With that said, this does not solve the root issue that we don't have a way of receiving feedback/understanding if a pod is receiving too many RPS. The current proposal in that PR simply tries to balance subsets so that each subset has a number of endpoints that is proportional to the number of nodes that zone/subset includes. There is _lots_ of room for improvement there, but I think the proposal is a step in the right direction and sufficiently generic that it enables future improvements.

Reply:
My reading of this proposal is the introduction of some useful primitives to allow topology aware slice subsetting, but it's really up to the controller to make good choices with those primitives. I like this approach of separating subsetting logic (#2005) from consumption, but I think this KEP could use a short description of a toy controller and what could happen with a few example services of various sizes and topologies - that may help highlight specific use cases.

Reply:
Thanks Jeremy - I think that's at least part of the thing that I was trying to say.

So let me try to clarify what I meant - I agree what I wrote isn't super clear.
Basically - I was NOT saying that we should describe here exactly how the subsetting algorithm will work - it's fine to do that in a separate KEP later.
But I would like to ensure that the approach proposed in this KEP (i.e. introducing these new labels), which from the first glance looks fine will not cause us significant issues later when we will start using it at scale.

So imagine the following scenario:
- I have 10 clusters, each of them in a different region and spanning 3 different zones (so 30 zones in total)
- I silently assumed that each kube-proxy will be watching EndpointSlices for its own zone and its own region - [but please make that very explicit in the proposal too if that's the case]
So then there are a couple questions:
- I'm trying to reason about zone vs region - initially that region will be sum over its zones, but then that doesn't make sense (as it will be supeset of zone which makes zone unneded). So it seems it has to be somehow independent to be somehow useful. Is that right?
- I'm also trying to thing about fan-out. Let's say I have a very small service (say 3 endpoints). And I'm upgrading one of its endpoints. For small case, subsetting probably doesn't make sense. So does that mean that will have to update EndpointSlice for every single zone now? So 30 of objects? That seems huuge.
But even if we focus on single cluster and the fact that majority of services are very small, the 3 fan-out factor for writes sounds quite high. I would really like to understand if there are any plans to address this (if there aren't I'm not sure it's really the approach we want). Or maybe I don't understand it?

Reply:
Thanks for the great feedback here!

> I'm trying to reason about zone vs region - initially that region will be sum over its zones, but then that doesn't make sense (as it will be supeset of zone which makes zone unneded). So it seems it has to be somehow independent to be somehow useful. Is that right?

In #2005 the proposal is entirely covering Zones, and `for-zone` labels could be added to EndpointSlices to denote that they should only be consumed by EndpointSlices in that zone. The KEP briefly mentions that the same pattern could apply to region. It's unclear if we would ever want to set both a zone and region label on the same EndpointSlice. I think most use cases could be accomplished by setting only one:

`RequireZone`, `PreferZone`, `PreferZonePreferRegion`, `PreferZoneRequireRegion`:
- Only a `for-zone` label is needed. In any case where traffic is required to stay in region, the controller could simply take that into account for what endpoints were delivered to a zone.

`PreferRegion`, `RequireRegion`:
- Only a `for-region` label is needed.


>I'm also trying to thing about fan-out. Let's say I have a very small service (say 3 endpoints). And I'm upgrading one of its endpoints. For small case, subsetting probably doesn't make sense. So does that mean that will have to update EndpointSlice for every single zone now? So 30 of objects? That seems huuge.
But even if we focus on single cluster and the fact that majority of services are very small, the 3 fan-out factor for writes sounds quite high. I would really like to understand if there are any plans to address this (if there aren't I'm not sure it's really the approach we want). Or maybe I don't understand it?

Beyond just backwards compatibility, this is an important reason why we need to continue to consume all EndpointSlices without a `for-zone` or `for-region` label. There will be some cases where it's simply not practical to do subsetting because the scale is too small.

In #2005 the proposal for any `Prefer` approach involves a starting threshold. Below that threshold we continue to deliver all EndpointSlices to all consumers with these unlabeled EndpointSlices. This essentially means that any kind of Topology Aware Routing or Subsetting approach would not kick in until Services had a sufficient number of Endpoints. Based on our simulations, setting this threshold at 3x the number of zones seemed to provide a reasonable balance.

We don't have the same concept for any `Require` based approach though and in that case that would dramatically increase the number of EndpointSlices for a Service, especially when a cluster/environment has a large number of zones. 

With that said, this example was a multicluster scenario, and everything I've been talking about is how the EndpointSlice controller would behave in a single cluster. I imagine @JeremyOT would have a better idea of how a controller might subset endpoints across a multicluster environment.

Reply:
Re zone vs region:
- I think I was trying to get even further. I was wondering what the "EndpointSlice" with "for-region" would actually be. Wouldn't that effectively be a "sum over zones from that region of EndpointSlices with "for-zone""? That immediately means that every pod change would result in two endpoint slices updated (and effectively 2x data duplication)?

- Re fan-out: I think that 2x is inevitable then, right? Because we need to create the ones without any labels (for backward compatibility), and then each endpoint will probably be part of at least one EndpointSlice with "for-zone" label?

[I'm not necessary saying that we don't want to take this approach. But a bunch of things mentioned in this comment thread should actually be part of the KEP itself to allow more conscious decision.]

Reply:
Oh I think I get the confusion now. None of the proposed approaches would involve data duplication. The reason they might end up with more EndpointSlices would be less efficient packing. Here's the number of EndpointSlices that would result based on the number of endpoints a Service has in a 3 zone cluster.

| # endpoints | Original # slices | PreferZone # slices | RequireZone # slices |
|-|-|-|-|
| 6 | 1 | 1 | 3 |
| 80 | 1 | 3 | 3 |
| 250 | 3 | 3 | 3 |

The RequireZone approach requires at least one EndpointSlice per zone per Service. The PreferZone also has the same requirement unless the minimum threshold has been met. Before that threshold, a single shared EndpointSlice (no additional labels) is used. There's some padding involved here to make sure we're not flapping back and forth between these states.

With this approach EndpointSlices can be delivered everywhere (no additional labels), or to a zone (for-zone), or to a region (for-region). None of the proposed approaches involve a single Service having separate sets of EndpointSlices for each use case.

As far as backwards compatibility, we don't need create EndpointSlices without labels for that, we just need to ensure that consumers always support consuming EndpointSlices without these labels. Even if we updated the EndpointSlice controller to consistently label these EndpointSlices with `for-zone` or `for-region`, we couldn't guarantee that other producers would. 

With that said, there's nothing in any current consumer implementation that would break if additional labels like `for-zone` or `for-region` were added to EndpointSlices. All consumers will need to care about with both this and the original approach is the `kubernetes.io/service-name` label. If they want to support subsetting, they can update their selectors as described in this KEP, but subsetting won't actually break any existing functionality. 

I'll have to update both KEPs to make this all more clear, thanks for the great feedback!

Reply:
OK - that makes more sense now. What you're basically saying is that even in "prefer zone" case, we will not send all the endpoints to all kube-proxies. Which I need to think about more a little bit (in the context of my original doc I shared you yesterday).

That said - all that stuff should be added to the KEP - if I didn't understand this, I guess others may not understand that too.

Reply:
Good call - I've added all of this information to the KEP so it's more clear.